### PR TITLE
Support json format output with `outdated` command

### PR DESCRIPTION
**Pull request by Shinyaigeek**
## Summary

Support json format output with `outdated` command

## Ticket

Closes: #2705

## Details

support JSON format output with `outdated` command if `--format json` is specified. 

## Test

I tested this with my own project,

```
pnpm outdated --format json
```

```json
{
        "@types/yargs": {
                "current": "17.0.11",
                "latest": "17,0.13",
                "dependencyKind": "dev"
        },
        "dotenv": {
                "current": "16.0.1",
                "latest": "16,0.3",
                "dependencyKind": "dev"
        },
        "jsdom": {
                "current": "20.0.0",
                "latest": "20,0.2",
                "dependencyKind": "dev"
        },
        "tslib": {
                "current": "2.4.0",
                "latest": "2,4.1"
        }
}
```

### CODE REVIEW

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 1]
```diff
@@ -39,6 +39,7 @@ export function rcOptionsTypes () {
     ], allTypes),
     compatible: Boolean,
     table: Boolean,
+    format: ['json'],
```
Comment (Shinyaigeek) [inline packages/plugin-commands-outdated/src/outdated.ts]: If we support this feature, it may be better to merge `table` and `format` options. such as `format: ['json', 'table', 'list']`

**Comment by zkochan** [comment]
Sounds good

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 2]
```diff
@@ -244,6 +250,37 @@ ${renderCurrent(outdatedPkg)} ${chalk.grey('=>')} ${renderLatest(outdatedPkg)}`
     .join('\n\n') + '\n'
 }
 
+function renderOutdatedJSON (outdatedPackages: readonly OutdatedPackage[], opts: { long?: boolean }) {
+  const outdatedPackagesJSON = {} as {
```
Comment (Shinyaigeek) [inline packages/plugin-commands-outdated/src/outdated.ts]: I wondered whether I should create a function that is almost equivalent to `renderLatest`, only without the output with chalk, but I was concerned that the implementation would be scattered, so I implemented it by adding `withoutChalk` argument which disables chalk output

**Comment by Shinyaigeek** [comment]
@zkochan first implementation is done!, so please review this when you are free 🙇

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 1 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: adding this, I also think this would be good to support `markdown` table format with `outdated` command. I would like to use `outdated` command's output with GitHub's issue, and it will be helpful to support `markdown` table format

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 3]
```diff
@@ -289,26 +326,32 @@ export function renderCurrent ({ current, wanted }: OutdatedPackage) {
   return `${output} (wanted ${wanted})`
 }
 
-export function renderLatest (outdatedPkg: OutdatedWithVersionDiff): string {
+export function renderLatest (outdatedPkg: OutdatedWithVersionDiff, withoutChalk = false): string {
```
Comment (zkochan) [inline packages/plugin-commands-outdated/src/outdated.ts]: I don't see any benefit from using this function. Just create a new one.

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 4]
```diff
@@ -289,26 +326,32 @@ export function renderCurrent ({ current, wanted }: OutdatedPackage) {
...
 }
 
-export function renderDetails ({ latestManifest }: OutdatedPackage) {
+export function renderDetails ({ latestManifest }: OutdatedPackage, withoutChalk = false) {
   if (latestManifest == null) return ''
   const outputs = []
   if (latestManifest.deprecated) {
-    outputs.push(wrapAnsi(chalk.redBright(latestManifest.deprecated), 40))
+    const detail = latestManifest.deprecated
+    outputs.push(wrapAnsi(withoutChalk ? detail : chalk.redBright(detail), 40))
```
Comment (zkochan) [inline packages/plugin-commands-outdated/src/outdated.ts]: why do you need to wrap for the json output?

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 5]
```diff
@@ -289,26 +326,32 @@ export function renderCurrent ({ current, wanted }: OutdatedPackage) {
...
 
+  if (withoutChalk) {
+    return diff.flat().join('.')
+  }
+
   return colorizeSemverDiff({ change, diff })
 }
 
-export function renderDetails ({ latestManifest }: OutdatedPackage) {
+export function renderDetails ({ latestManifest }: OutdatedPackage, withoutChalk = false) {
```
Comment (zkochan) [inline packages/plugin-commands-outdated/src/outdated.ts]: create a new function for json

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 6]
```diff
@@ -244,6 +250,37 @@ ${renderCurrent(outdatedPkg)} ${chalk.grey('=>')} ${renderLatest(outdatedPkg)}`
     .join('\n\n') + '\n'
 }
 
+function renderOutdatedJSON (outdatedPackages: readonly OutdatedPackage[], opts: { long?: boolean }) {
+  const outdatedPackagesJSON = {} as {
+    [outdatedPackageName: string]: {
+      current: string
+      latest: string
+      dependencyKind?: 'dev' | 'optional'
```
Comment (zkochan) [inline packages/plugin-commands-outdated/src/outdated.ts]: why is belongsTo not good? 

if you want to keep it, rename it to `dependencyType` and probably make it required.

File: packages/plugin-commands-outdated/src/recursive.ts [hunk 7]
```diff
@@ -153,6 +157,39 @@ ${renderCurrent(outdatedPkg)} ${chalk.grey('=>')} ${renderLatest(outdatedPkg)}`
     .join('\n\n') + '\n'
 }
 
+function renderOutdatedJSON (outdatedMap: Record<string, OutdatedInWorkspace>, opts: { long?: boolean }) {
```
Comment (zkochan) [inline packages/plugin-commands-outdated/src/recursive.ts]: why isn't the similar function from outdated.ts reused here?

File: packages/plugin-commands-outdated/test/index.ts [hunk 8]
```diff
@@ -190,6 +190,65 @@ https://github.com/kevva/is-positive#readme
...
+\t"is-negative": {
+\t\t"current": "1.0.0",
+\t\t"latest": "2.1.0"
+\t},
+\t"is-positive": {
+\t\t"current": "1.0.0",
+\t\t"latest": "3.1.0",
+\t\t"dependencyKind": "dev"
+\t}
+}`)
```
Comment (zkochan) [inline packages/plugin-commands-outdated/test/index.ts]: it looks pretty ugly. You could just run `JSON.stringify` to generate the expected output.

File: packages/plugin-commands-outdated/test/index.ts [hunk 9]
```diff
@@ -190,6 +190,65 @@ https://github.com/kevva/is-positive#readme
...
+https://foo.bar/qar
+
+is-negative
+1.0.0 => 2.1.0
+https://github.com/kevva/is-negative#readme
+
+is-positive (dev)
+1.0.0 => 3.1.0
+https://github.com/kevva/is-positive#readme
+`)
```
Comment (zkochan) [inline packages/plugin-commands-outdated/test/index.ts]: why is this needed? This format already existed before your changes. Why do you add a test for it?

**Review by zkochan — CHANGES_REQUESTED** [review]

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 3 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: I was concerned about a scattering of implementation, but this is not a firm intention, so fixed in [90069a4](https://github.com/pnpm/pnpm/pull/5582/commits/90069a4fa1545f84df758f765869f02c31eed3e2) 👍

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 5 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: same here https://github.com/pnpm/pnpm/pull/5582#discussion_r1014605452

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 4 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: Sorry, just mistake. [90069a4](https://github.com/pnpm/pnpm/pull/5582/commits/90069a4fa1545f84df758f765869f02c31eed3e2) :+1:

File: packages/plugin-commands-outdated/test/index.ts (hunk 9 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/test/index.ts]: Sorry, just mistake. [c92fd62](https://github.com/pnpm/pnpm/pull/5582/commits/c92fd62d66ceb9097e2311e0e80c4f3499fec38c) :+1:

File: packages/plugin-commands-outdated/test/index.ts (hunk 8 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/test/index.ts]: good idea! Thank you! [2f3f4f5](https://github.com/pnpm/pnpm/pull/5582/commits/2f3f4f5e82ce83f1ce24db2ca551862cc6977ff1) 👍

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 6 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: https://github.com/pnpm/pnpm/blob/main/packages/plugin-commands-outdated/src/outdated.ts#L278 I was going to fit to `renderPackageName`, but this is not a firm intention. fixed in 5abaceb

File: packages/plugin-commands-outdated/src/recursive.ts (hunk 7 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/recursive.ts]: I was wondering which or not I should re-implement `renderOutdatedJSON` also in `recursive.ts` or not. I decided to re-implement it because

- in recursive, `dependentPackages` field is added and this requires `OutdatedInWorkspace` (but this is trivial)
- `sortOutdatedPackages` function, which is used in both of `outdated` and recursive, is a bit different in `outdated` and recursive…

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 1 above)
Comment (zkochan) [reply packages/plugin-commands-outdated/src/outdated.ts]: > If we support this feature, it may be better to merge table and format options. such as format: ['json', 'table', 'list']

sounds good. Regarding markdown, I am not sure how frequently that would be needed.

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 10]
```diff
@@ -244,6 +251,31 @@ ${renderCurrent(outdatedPkg)} ${chalk.grey('=>')} ${renderLatest(outdatedPkg)}`
...
+      current: string
+      latest: string
+      dependencyType: DependenciesField
+      details?: string
+    }
+  }
+  for (const outdatedPkg of sortOutdatedPackages(outdatedPackages)) {
+    const { packageName, belongsTo } = outdatedPkg
+    outdatedPackagesJSON[packageName] = {
+      current: renderCurrent(outdatedPkg),
```
Comment (zkochan) [inline packages/plugin-commands-outdated/src/outdated.ts]: It is a JSON, so I would rather return the unprocessed fields as they are. Like:

```suggestion
      current: outdatedPkg.current,
      wanted: outdatedPkg.wanted,
```

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 11]
```diff
@@ -244,6 +251,31 @@ ${renderCurrent(outdatedPkg)} ${chalk.grey('=>')} ${renderLatest(outdatedPkg)}`
...
+      latest: extractLatest(outdatedPkg),
```
Comment (zkochan) [inline packages/plugin-commands-outdated/src/outdated.ts]: I don't think you need a function. Just return
```suggestion
      latestVersion: outdatedPkg.latestManifest?.version,
      latestDeprecated: outdatedPkg.latestManifest?.deprecated,
```

File: packages/plugin-commands-outdated/src/outdated.ts [hunk 12]
```diff
@@ -244,6 +251,31 @@ ${renderCurrent(outdatedPkg)} ${chalk.grey('=>')} ${renderLatest(outdatedPkg)}`
...
+      dependencyType: belongsTo,
+    }
+
+    if (opts.long) {
+      outdatedPackagesJSON[packageName].details = extractDetails(outdatedPkg)
```
Comment (zkochan) [inline packages/plugin-commands-outdated/src/outdated.ts]: I think for long we can just pass the whole latest manifest
```suggestion
      outdatedPackagesJSON[packageName].latestManifest = outdatedPkg.latestManifest
```

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 10 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: Certainly!! This is about JSON format and this is usually used with automation process, so it is better to simplify. [2299772](https://github.com/pnpm/pnpm/pull/5582/commits/2299772b0d1dbac333acbda86ee56ac339eefc8d) :+1:

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 11 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: same as https://github.com/pnpm/pnpm/pull/5582#discussion_r1014654608

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 12 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: same as https://github.com/pnpm/pnpm/pull/5582#discussion_r1014654608

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 11 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: I thought user want to know which package is deprecated, but they only know which package is deprecated by checking `current` field is equal to `latestVersion` field and this is not friendly. So I added `deprecated` field.

File: packages/plugin-commands-outdated/src/outdated.ts (hunk 1 above)
Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: I added the process to handle `table` and `list` format option in 5d82338. Should we deprecate `table` option in `outdated` command?

Comment (Shinyaigeek) [reply packages/plugin-commands-outdated/src/outdated.ts]: As for the `markdown` table format, the only use case I can come up with is copying and pasting the output into a GitHub issue. I feel it is good to support this, but should we support this? (Of course this will be implemented in the other PR if we will do)

**Review by zkochan — APPROVED** [review]
